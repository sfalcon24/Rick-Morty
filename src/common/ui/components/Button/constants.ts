import {colors} from 'common/ui/theme/colors';
import type {ButtonVariant} from './types';

export const backgroundColor: Record<ButtonVariant, string> = {
  primary: colors.blue,
  secondary: colors.yellow,
};

export const textColor: Record<ButtonVariant, string> = {
  primary: colors.white,
  secondary: colors.black,
};
