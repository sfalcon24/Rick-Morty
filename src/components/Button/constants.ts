import {colors} from 'theme/colors';
import type {ButtonVariant} from './types';

export const backgroundColor: Record<ButtonVariant, string> = {
  primary: colors.blue,
  secondary: colors.yellow,
};

export const textColor: Record<ButtonVariant, string> = {
  primary: colors.white,
  secondary: colors.black,
};
