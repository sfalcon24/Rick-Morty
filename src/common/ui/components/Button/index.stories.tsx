import {action} from '@storybook/addon-actions';
import Button from '.';

export default {
  title: 'Button',
  component: Button,
};

export const Primary = () => (
  <Button variant="primary" title="Primary" onPress={action('onPress')} />
);

export const PrimaryLoading = () => (
  <Button
    variant="primary"
    title="Primary loading"
    onPress={action('onPress')}
    loading
  />
);

export const Secondary = () => (
  <Button variant="secondary" title="Secondary" onPress={action('onPress')} />
);

export const SecondaryLoading = () => (
  <Button
    variant="secondary"
    title="Secondary loading"
    onPress={action('onPress')}
    loading
  />
);
