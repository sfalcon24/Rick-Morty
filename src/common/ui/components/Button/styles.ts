import styled from 'styled-components/native';
import Text from 'common/ui/components/Text';
import {RectButton} from 'react-native-gesture-handler';
import {BACKGROUND_COLOR, TEXT_COLOR, WIDTH} from './constants';
import type {ContainerProps, TitleProps} from './types';

export const Container = styled.View<ContainerProps>`
  background-color: ${({variant}) => BACKGROUND_COLOR[variant]};
  border-width: ${({variant}) => (variant === 'large-dark' ? 2 : 0)}px;
  border-color: ${({theme}) => theme.colors.adding};
  width: ${({variant}) => WIDTH[variant]}px;
  height: ${({variant}) =>
    variant === 'large-dark' || variant === 'large'
      ? '50px'
      : variant === 'small'
      ? '28px'
      : '42px'};
  left: ${({variant}) =>
    variant === 'large-dark' || variant === 'large'
      ? '26px'
      : variant === 'small'
      ? '107px'
      : '0px'};
  top: ${({variant}) =>
    variant === 'large-dark' || variant === 'large'
      ? '100px'
      : variant === 'small'
      ? '8px'
      : '0px'};
  gap: ${({variant}) => (variant === 'small' ? 10 : 0)}px;
  padding: ${({variant}) => (variant === 'small' ? '0px 12px' : '0px')};
  border-radius: 16px;
`;

export const Pressable = styled(RectButton)`
  align-items: center;
  border-radius: 16px;
  height: 48px;
  justify-content: center;
`;

export const Title = styled(Text).attrs<TitleProps>(({buttonVariant}) => ({
  variant: buttonVariant === 'small' ? 'tagline13-tag' : 'body17-semibold',
}))<TitleProps>`
  color: ${({buttonVariant}) => TEXT_COLOR[buttonVariant]};
`;
