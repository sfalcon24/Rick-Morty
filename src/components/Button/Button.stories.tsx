import {Alert} from 'react-native';
import Button from '.';

export default {
  title: 'Button',
  component: Button,
};

const showAlert = () => {
  Alert.alert('Button pressed');
};

export const Primary = () => (
  <Button variant="primary" title="Primary" onPress={showAlert} />
);

export const PrimaryLoading = () => (
  <Button
    variant="primary"
    title="Primary loading"
    onPress={showAlert}
    loading
  />
);

export const Secondary = () => (
  <Button variant="secondary" title="Secondary" onPress={showAlert} />
);

export const SecondaryLoading = () => (
  <Button
    variant="secondary"
    title="Secondary loading"
    onPress={showAlert}
    loading
  />
);
