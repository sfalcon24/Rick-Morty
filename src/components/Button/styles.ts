import styled from 'styled-components/native';
import Text from 'components/Text';
import {RectButton} from 'react-native-gesture-handler';
import {backgroundColor, textColor} from './constants';
import type {ContainerProps, TextProps} from './types';

export const Container = styled.View<ContainerProps>`
  background-color: ${({variant}) => backgroundColor[variant]};
  border-radius: 16px;
`;

export const Pressable = styled(RectButton)`
  align-items: center;
  border-radius: 16px;
  height: 48px;
  justify-content: center;
`;

export const Title = styled(Text)<TextProps>`
  color: ${({variant}) => textColor[variant]};
`;
