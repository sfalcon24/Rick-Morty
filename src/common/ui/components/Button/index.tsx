import type {FC} from 'react';
import {memo} from 'react';
import React from 'react';
import {Container, Pressable, Title} from './styles';
import type {Props} from './types';

const Button: FC<Props> = ({
  loading,
  onPress,
  style,
  testID,
  title,
  variant = 'primary',
}) => (
  <Container style={style} variant={variant}>
    <Pressable
      enabled={!loading && !!onPress}
      onPress={onPress}
      testID={testID}>
      <Title variant={variant}>{title}</Title>
    </Pressable>
  </Container>
);

export default memo(Button);
