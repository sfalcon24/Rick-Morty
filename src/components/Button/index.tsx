import type {FC} from 'react';
import {memo} from 'react';
import React from 'react';
import {Container, Pressable, Title} from './styles';
import type {Props} from './types';

const Button: FC<Props> = ({
  loading,
  onPress,
  style,
  title,
  variant = 'primary',
}) => (
  <Container style={style} variant={variant}>
    <Pressable onPress={onPress} enabled={!loading && !!onPress}>
      <Title variant={variant}>{title}</Title>
    </Pressable>
  </Container>
);

export default memo(Button);
