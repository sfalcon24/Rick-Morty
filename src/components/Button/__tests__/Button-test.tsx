import 'react-native';
import renderer from 'react-test-renderer';
import Button from '../';

describe('Button', () => {
  it('Primary button renders', () => {
    const tree = renderer
      .create(<Button title="Test" variant="primary" />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Secondary button renders', () => {
    const tree = renderer
      .create(<Button title="Test" variant="secondary" />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
