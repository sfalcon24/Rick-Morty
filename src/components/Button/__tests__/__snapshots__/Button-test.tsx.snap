// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Primary button renders 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#0000FF",
        "borderRadius": 16,
      },
    ]
  }
  variant="primary"
>
  <RNGestureHandlerButton
    activeOpacity={0.105}
    collapsable={false}
    enabled={false}
    handlerTag={1}
    handlerType="NativeViewGestureHandler"
    onActiveStateChange={[Function]}
    onGestureEvent={[Function]}
    onGestureHandlerEvent={[Function]}
    onGestureHandlerStateChange={[Function]}
    onHandlerStateChange={[Function]}
    style={
      Object {
        "alignItems": "center",
        "borderRadius": 16,
        "height": 48,
        "justifyContent": "center",
      }
    }
    underlayColor="black"
  >
    <View
      collapsable={false}
      style={
        Object {
          "backgroundColor": "black",
          "borderBottomLeftRadius": undefined,
          "borderBottomRightRadius": undefined,
          "borderRadius": 16,
          "borderTopLeftRadius": undefined,
          "borderTopRightRadius": undefined,
          "bottom": 0,
          "left": 0,
          "opacity": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    />
    <Text
      style={
        Array [
          Object {
            "fontFamily": undefined,
          },
          Array [
            Object {
              "color": "#FFFFFF",
            },
          ],
        ]
      }
      variant="primary"
    >
      Test
    </Text>
  </RNGestureHandlerButton>
</View>
`;

exports[`Button Secondary button renders 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#FFFF00",
        "borderRadius": 16,
      },
    ]
  }
  variant="secondary"
>
  <RNGestureHandlerButton
    activeOpacity={0.105}
    collapsable={false}
    enabled={false}
    handlerTag={2}
    handlerType="NativeViewGestureHandler"
    onActiveStateChange={[Function]}
    onGestureEvent={[Function]}
    onGestureHandlerEvent={[Function]}
    onGestureHandlerStateChange={[Function]}
    onHandlerStateChange={[Function]}
    style={
      Object {
        "alignItems": "center",
        "borderRadius": 16,
        "height": 48,
        "justifyContent": "center",
      }
    }
    underlayColor="black"
  >
    <View
      collapsable={false}
      style={
        Object {
          "backgroundColor": "black",
          "borderBottomLeftRadius": undefined,
          "borderBottomRightRadius": undefined,
          "borderRadius": 16,
          "borderTopLeftRadius": undefined,
          "borderTopRightRadius": undefined,
          "bottom": 0,
          "left": 0,
          "opacity": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    />
    <Text
      style={
        Array [
          Object {
            "fontFamily": undefined,
          },
          Array [
            Object {
              "color": "#000000",
            },
          ],
        ]
      }
      variant="secondary"
    >
      Test
    </Text>
  </RNGestureHandlerButton>
</View>
`;
